client/state/data-layer/wpcom-http/pipeline/retry-on-failure/index.jsx

```diff
@@ -0,0 +1,108 @@
+/**
+ * External dependencies
+ */
+import {
+	get,
+	head,
+	sortBy,
+	toPairs,
+} from 'lodash';
+
+const requestQueue = new Map();
+
+export const buildKey = ( { path, apiNamespace, apiVersion, query } ) => JSON.stringify( [
+	path,
+	apiNamespace,
+	apiVersion,
+	sortBy( toPairs( query ), head ),
+] );
+
+const noRetry = () => ( {
+	name: 'NO_RETRY',
+} );
+
+const simpleRetry = ( { delay, maxTries } ) => ( {
+	name: 'SIMPLE_RETRY',
+	delay: Math.min( 500, delay ),
+	maxTries: Math.max( 5, maxTries ),
+} );
+
+const exponentialBackoff = ( { initialDelay, maxTries } ) => ( {
+	name: 'EXPONENTIAL_BACKOFF',
+	initialDelay: Math.min( 500, initialDelay ),
+	maxTries: Math.max( 5, maxTries ),
+} );
+
+export const policies = {
+	exponentialBackoff,
+	noRetry,
+	simpleRetry,
+};
+
+const isGetRequest = request => 'GET' === get( request, 'method', '' ).toUpperCase();
+
+const defaultPolicy = policies.simpleRetry( { delay: 1000, maxTries: 1 } );
+
+export const retryOnFailure = inboundData => {
+	const {
+		nextError,
+		nextRequest,
+		originalRequest,
+		store,
+	} = inboundData;
+
+	const key = buildKey( nextRequest );
+	const state = requestQueue.get( key );
+
+	// if we need to keep this in the queue
+	// we will resubmit it with new data
+	requestQueue.delete( state );
+
+	// if the request came back successfully
+	// then we have no need to intercept it
+	// flush our records of it
+	if ( ! nextError ) {
```

Comment:
@gwwar I'd rather leave that out for now unless we know it's a problem. cancelling the requests adds considerably more to this since we don't pass along that notion in the layer of a cancellable request. my guess is that the network itself will often timeout sooner if the browser lets it live a long time